### PR TITLE
Multiplayer improvements

### LINKED ISSUE #130: Display the score for all Landers in that session (co-op)

### LINKED ISSUE #191: Pressing R will delete all clients when it shouldn't

### LINKED ISSUE #192: Client connecting to host has delay. See comments.
Client takes a long time to connect. Need a status message or something on client so they have feedback.

### LINKED ISSUE #193: Need to present host with message when client joins

### LINKED ISSUE #227: Don't send MASS data over enet. See comments.
Try sending a skinny version of the lander table across enet and let the client determine draw as much as possible on the client side.

### PR DESCRIPTION
**Pull request by togfoxy**
Also includes some efficiencies in functions.ResetGame()

Changes proposed in this pull request.
- host is notified when a client connects
- client is notified when it tries to connect
- host now sends a skinny version of lander to reduce net lag
- some efficiencies in ResetGame
- the score for every lander is now displayed (relevant in mplayer mode)
- introduced a new keypress: the return/enter/keypad enter key now resets just the one lander. The 'r' key is untouched and resets the whole game for every lander (mplayer mode)
- introduced a Lander.reset function that resets a single lander
- each lander's score is now saved in the lander table and refreshed during Lander.update

Fixes #191 
Fixes #192 
Fixes #193 
Fixes #130 
Fixes #227 

[untested. still WIP and draft only. Not ready for review]

### CODE REVIEW

File: source/hud.lua [hunk 1]
```diff
@@ -171,14 +171,20 @@ end
...
+	local alignment	= "right"
+	
 	Assets.setFont("font14")
-	love.graphics.printf("Score: " .. score, 0, 75, SCREEN_WIDTH, "center")
-	love.graphics.printf("High Score: " .. highScore, 0, 90, SCREEN_WIDTH, "center")
+	for _,lander in pairs(LANDERS) do
+		local roundedScore = Cf.round(lander.score)
+		local formattedScore = Cf.strFormatThousand(roundedScore)		
+		local tempString = lander.name .. ": " .. formattedScore
+		love.graphics.printf(tempString,x,y, lineLength, alignment)
```
Comment (MadByteDE) [inline source/hud.lua:185]: I'd just use `text` instead of `tempString`

**Comment by MadByteDE** [comment]
Some nice fixes in here!
Left some comments for the draft changes.

**Comment by togfoxy** [comment]
Added some neat features. Host and client are the player names. IRL they would be things like mE_FLi_fasTEr and **ZZOM ZZOM**

8-)

Please review for obvious errors.

Untitled

**Review by togfoxy — COMMENTED** [review]
Really ready. For real. I think.

**Comment by MadByteDE** [comment]
Right now your PR would result in the player not being able to play SP or MP at all.
The reason is that you moved `Terrain.init()` from `Fun.ResetGame` into `love.load`, right before calling `Fun.ResetGame()` which basically means you call stuff in this order:
```lua
	Terrain.init()
	OBJECTS = {}
	LANDERS = {}
	table.insert(LANDERS, Lander.create())
```
It basically means you generate new terrain **and objects** with `Terrain.init` but then delete the objects again with `OBJECTS = {}`. And because the lander depends on at least one base being generated, the game will show an error, no matter if you start a new game with `New Game` or `Host Game`. And if no save file exists, `Load Game` also shows the same error.

Also, out of curiosity, why did you remove `Smoke.destroy` from `ResetGame`? It's needed to delete smoke particles between sessions.

If your intention was to use `ResetGame` to just reset the landers while keeping the objects and ground, I'd suggest to add a new function that only resets the landers.

Another issue is that if a player in a running MP sessions starts a `New Game` or loads a game from the menu, the old session is not reset correctly. In both instances (client and server) a third player will be spawned at the originX and the old landers remain at their previous locations. 

And last but not least, if the host starts a new game, his lander get's stuck in place and can't move anymore.

I'm really sorry to say this, but I believe you need to revisit your code again :disappointed: 

Here is a video showing the behavior:

https://user-images.githubusercontent.com/4752038/141341852-6a42afd2-d0a7-457d-a123-8788ab8b0c91.mp4

**Comment by togfoxy** [comment]
I actually get a different runtime bug. My local branches must be tangled.

This is a non-essential PR so I'll close with no merge so it doesn't block other good work. Thanks.

**Pull request closed without merging**